### PR TITLE
roachtest: add ability to set default virtual cluster

### PR DESCRIPTION
**Pull request by renatolabs**
This commit adds a `SetDefaultVirtualCluster` function to the cluster interface. This function's primary use case is to simulate the behaviour of the `server.controller.default_target_cluster` cluster setting, but for separate-process tenants. When this function is called, every subsequent call to `Conn` will connect to the provided virtual cluster (instead of `system`). Similary, expansions provided by roachprod (such as `{pgurl}`) will also use the new default virtual cluster. This functionality will enable the `mixedversion` framework to continue to provide the existing API when testing separate-process tenant upgrades. It could also be useful if roachtest itself wanted to provide this kind of randomized deployment model in each run.

Epic: none

Release note: None

### CODE REVIEW

**Review by srosenberg — DISMISSED** [review]

**Comment by renatolabs** [comment]
Requesting another review pass on this PR.

I wanted to change `ExternalAdminUIAddr` to be able to take an optional "virtual cluster" parameter (it would hardcode system previously). However, that turned out to hit a wall: we already have `option.VirtualClusterName` defined in the roachtest `option` package, but that function is defined as an option for `c.Conn{,E}`. Ideally, we would use the same `option.VirtualClusterName` API for the `ExernalAdminUIAddr` function as well.

A couple of approaches we could consider:

* have separate option functions for each cluster function: so we could have `option.AdminAddrVirtualCluster`, which is both awkward and error prone (why can't I use `option.VirtualCluster`?)
* give up on the "functional options" pattern and make `virtualCluster` a required parameter to `ExternalAdminUIAddr`. This could work, but it's inconvenient to have to pass `install.SystemInterfaceName` to every call of that function, even in cases where there's no multi tenancy at all.

I pushed a commit that uses some reflection to achieve a more ergonomic API: `option.VirtualClusterName` can now be used for both `c.Conn` and `c.ExternalAdminUIAddr`. Reflection always (rightfully) raises some eyebrows but I think in this case it's the lesser of the evils.

Open to thoughts and if anyone thinks this should be handled in a different way, I'm all ears.

File: pkg/cmd/roachtest/option/cluster_options.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
+// Copyright 2023 The Cockroach Authors.
```
Comment (srosenberg) [inline pkg/cmd/roachtest/option/cluster_options.go:1]: Nit: is it already 2024? I always forget :)

File: pkg/cmd/roachtest/option/cluster_options.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+				retErr = fmt.Errorf("failed to set %q on %T: %v", currentField, container, r)
+			}
+		}
+	}()
+
+	for _, opt := range opts {
+		currentField = opt.name
+		f := s.Elem().FieldByName(currentField)
+		if !f.IsValid() {
+			return fmt.Errorf("invalid option %s for %T", opt.name, container)
```
Comment (srosenberg) [inline pkg/cmd/roachtest/option/cluster_options.go:120]: Nit: `%q` for consistency? (Above uses `%q` for `currentField`.)

**Comment by srosenberg** [comment]
> However, that turned out to hit a wall: we already have `option.VirtualClusterName` defined in the roachtest `option` package, but that function is defined as an option for `c.Conn{,E}`.

Could we just pretend that `ConnOption` is a _union_ type and pass it to `ExternalAdminUIAddr`? i.e.,

```
ExternalAdminUIAddr(ctx context.Context, l *logger.Logger, node option.NodeListOption, opts ...func(*option.ConnOption)) ([]string, error)
```

and call `c.ExternalAdminUIAddr(ctx, l, nodes, option.VirtualClusterName("foo"))`. I have a feeling I'm missing something...

**Comment by renatolabs** [comment]
> Couuld we just pretend that `ConnOption` is a union type and pass it to `ExternalAdminUIAddr`?

We could do something like this. I considered that option, but the reason I didn't like it as much is that then you give up on any kind of option validation -- one could pass `ExternalAdminUIAddr(..., option.AuthOption(...))` and the code would compile, but the option would be silently ignored because it doesn't apply. Dead code would be undetectable and confusing. It would also be harder to tell what options are applicable in each function that takes these options (assuming we will over time desire to have more functions that take options, which I think is a likely possibility.)

File: pkg/cmd/roachtest/option/cluster_options.go (hunk 1 above)
Comment (renatolabs) [reply pkg/cmd/roachtest/option/cluster_options.go:1]: Do we update copyright headers when files are moved?

Comment (srosenberg) [reply pkg/cmd/roachtest/option/cluster_options.go:1]: Nope, I don't think it's common practice to update copyright headers upon renaming. I actually didn't realize it was a rename; at the time of writing, I thought it was a newly added file. 2023 it is!

**Comment by srosenberg** [comment]
> We could do something like this. I considered that option, but the reason I didn't like it as much is that then you give up on any kind of option validation -- one could pass `ExternalAdminUIAddr(..., option.AuthOption(...))` and the code would compile, but the option would be silently ignored because it doesn't apply. Dead code would be undetectable and confusing. It would also be harder to tell what options are applicable in each function that takes these options (assuming we will over time desire to have more functions that take options, which I think is a likely possibility.)

Yep, that's indeed the problem with "duck typing" :) Lack of subtypes is at the root of the problem here. I am not opposed to your workaround, but it didn't strike me as being particularly intuitive; on the first reading, it took > 5 mins to "unparse" everything. Maybe other folks could comment on the pattern of `CustomOption`; my main worry is mixing different APIs of passing options. @DarrylWong @herkolategan Thoughts?

Note that we could also stick with `ConnOption` and use reflection, but only to _validate_ that the passed `opts ...func(*option.ConnOption)` doesn't set any fields outside of `option.VirtualClusterOptions`.

**Review by srosenberg — APPROVED** [review]

**Comment by DarrylWong** [comment]
I agree its a bit confusing but of the given solutions, `CustomOptions` approach seems fine to me.

>Note that we could also stick with ConnOption and use reflection, but only to validate

I like this solution more, but would this fall apart if we ever need to use these options in a non connection context? i.e. if we add  function that takes in `DBName` or `VirtualClusterName` but isn't a connection, then we're back to the same problem?

**Review by DarrylWong — APPROVED** [review]

**Comment by srosenberg** [comment]
> I like this solution more, but would this fall apart if we ever need to use these options in a non connection context? i.e. if we add function that takes in `DBName` or `VirtualClusterName` but isn't a connection, then we're back to the same problem?

Yep, here you can't overload `VirtualClusterName` constructor, so you're stuck with `ConnOption`, i.e.,

```
func VirtualClusterName(name string) func(*ConnOption) {
	return func(option *ConnOption) {
		option.VirtualClusterName = name
	}
}
```

Thus, anywhere you want to pass `VirtualClusterName`, you have to use `ConnOption` and validate. :(

**Comment by renatolabs** [comment]
> use reflection, but only to validate that the passed opts

I pushed a variation of the initial approach to implement something like this suggestion (although I suspect it might not have been exactly what you had in mind). The option definitions are a lot simpler and without the `applyFunc` business which is a big win. The validation function got a little more complex, but I don't think by much.

Preferences? I think I like this approach slightly more.

File: pkg/cmd/roachtest/option/cluster_options.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+		f := containerStruct.FieldByName(currentField)
+
+		// It is an error for the container struct to have fields that are
+		// not present in the `GlobalOptions` struct.
+		globalField, ok := globalFields[currentField]
+		if !ok {
+			return fmt.Errorf("options struct %T has unknown option %q", container, currentField)
+		}
+
+		f.Set(globalField)
```
Comment (srosenberg) [inline pkg/cmd/roachtest/option/cluster_options.go:151]: Oof, this is really the `apply` part, the rest is validation. Maybe worth a comment to help guide the reader?

**Comment by srosenberg** [comment]
> The option definitions are a lot simpler and without the `applyFunc` business which is a big win. 

I concur, definitely a big win; the user-side is a lot more intuitive now!

> The validation function got a little more complex, but I don't think by much.
> 

Yep. Not the prettiest function, but it does what needs to be done :)

> Preferences? I think I like this approach slightly more.

I'm definitely a lot more enthusiastic about this approach. Thanks more making it easier to grok for the mere-mortals like us! :)

File: pkg/cmd/roachtest/option/cluster_options.go (hunk 3 above)
Comment (renatolabs) [reply pkg/cmd/roachtest/option/cluster_options.go:153]: Done.

**Pull request merged**